---
title: "Drop-seq alignment report"
date: "`r format(Sys.time(), '%B %d, %Y')`"
params:
  rmd: "report.Rmd"
output:
  html_document:
    number_sections: yes
    toc: yes
---

<br>

```{r setup, echo = FALSE}
# set global chunk options
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_knit$set(root.dir = ".." )
```

This report shows summary statistics produced by the Drop-seq alignment pipeline for sample **`r snakemake@wildcards$sample`**.

# Cell and molecular barcodes
These tables shows the total number of detected cell and molecular (UMI) barcodes and how many barcodes had at least 1 failed bases. For full summary data see <tt>`r snakemake@input$cell_bcs`</tt> and <tt>`r snakemake@input$mol_bcs`</tt> files. Any reads with barcodes that have any failed bases were excluded from subsequent steps.

```{r}
# load required functions
source("scripts/barcode_summary.R")

# calculate summary statistics
bc_stats <- bc_summary(snakemake@input$cell_bcs, snakemake@input$mol_bcs)

# print table
knitr::kable(bc_stats)
```

***

# Read trimming
Sequenced reads might still contain parts of SMART adapters (5') and polyA tails (3'). The Drop-seq pipeline therefore scans reads and trims any remaining adapter or polyA tail sequences prior to mapping. These plots show the number of clipped bases or read lengths after clipping for all trimmed reads. For full summary data see <tt>`r snakemake@input$adapt_trim`</tt> and <tt>`r snakemake@input$polyA_trim`</tt> files.

<br>

```{r}
# load required functions
source("scripts/read_trimming_summary.R")

# extract total number of reads from STAR log output
input_reads  <- readLines(snakemake@input$star_smry, n = 6)[6]
total_reads <- as.numeric(unlist(strsplit(input_reads, split = "\t"))[2])

# plot adapter trimming summary
adapter_trim_hist(adapter_trim_file = snakemake@input$adapt_trim,
                  sample = snakemake@wildcards$sample,
                  total_reads = total_reads)
```

<br>

```{r}
# plot polyA trimming summary
polyA_trim_hist(polyA_trim_file = snakemake@input$polyA_trim,
                  sample = snakemake@wildcards$sample,
                  total_reads = total_reads)
```

***

# Read alignment
Reads were aligned to the provided reference genome using STAR. The following plots show a summary of the mapping results. For full summary see the <tt> `r snakemake@input$star_smry` </tt> file.

<br>

```{r}
# load required functions
source("scripts/mapping_stats.R")

# extract percentage of reads across mapping cats
perc_reads <- get_mapping_cats(snakemake@input$star_smry)

# plot percentage of reads
plot_mapping_cats(mapping_cats = perc_reads,
                  sample = snakemake@wildcards$sample,
                  total_reads = total_reads)
```

***

# Barcode synthesis errors
DetectBeadSynthesisErrors from Drop-seq tools scans barcodes for synthesis errors and tries to correct incomplete cell barcodes or discards them if this is not possible. For details on the method refer to the [Drop-seq Alignment Cookbook](http://mccarrolllab.com/dropseq/).

<br>

**Summary of detected error types**
```{r}
# load required functions
source("scripts/synthesis_error_summary.R")

# read synthesis error summary data
synt_error_summary <- read_synthesis_error(snakemake@input$synthesis_error)

# transpose metrics data.frame for better layout in html output
metrics <- t(synt_error_summary$metrics)

# add ":" to rownames
rownames(metrics) <- paste0(rownames(metrics), ":")

# print metrics summary table in list format
knitr::kable(metrics, col.names = "")
```

<br>
<br>

**Synthesis error**
The most common detected errors are cell barcodes that did not undergo 12 rounds of splitting and pooling, and are therefore shorter. This manifests itself by detected fixed Ts at the end of the UMI sequence (see [Drop-seq Alignment Cookbook](http://mccarrolllab.com/dropseq/) for more details. This histogram shows the number of cells with detected fixed Ts and the base positions where the errors are deteced.

<br>

```{r}
# plot histogram of detected fixed Ts in UMIs
synthesis_error_hist(synt_error_summary, sample = snakemake@wildcards$sample,
                     cell_bc_length = 12, mol_bc_length = 8)
```

***

# Reads per cell barcode
The cumulative distribution of reads per cell barcode is created to assess the number cells in each experiment. The number of reads per cell barcode is calculated and barcodes are sorted by number of associated reads. Then the cumulative fraction of reads as a function of the number of cell barcodes is calculated and plotted.

<br>

```{r}
# load required functions
source("scripts/reads_cumfrac.R")

# get sample id
sample <- snakemake@wildcards$sample

# calculate and plot cumulative fraction of reads per cell barcode
rpc <- cumfrac_rpc(snakemake@input$reads_per_cell)

# define plot title
title <- paste0(sample, " (total reads: ", total_reads, ")")

# extract expected cell number from snakemake object
expect_cells <- as.numeric(snakemake@config$expect_cell_numbers[sample])

# plot cumulative fraction plot
plot_cumfrac_rpc(rpc, nbcs = expect_cells * 10, title = title)
```

***

# Source
* <a download="report.Rmd" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd', encoding = 'base64')`">R Markdown source file (to produce this document)</a>
* [Drop-seq webpage (McCarroll Lab)](http://mccarrolllab.com/dropseq/)